### PR TITLE
H-2611: Make Postgres query compiler more flexible

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

The current compiler is very restricted to be used in `SELECT` statements. Especially, `Column` also captures a path to a JSON property (e.g. for `EntityEditions::Properties(JsonField)`). Also `Column` transpiles to `"{table}"."{column}"` which is not valid for `INSERT`. This generalizes the query builder and use `Expression` where appropriate.

## 🚫 Blocked by

- #4362 

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change
### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: apps/hash-graph/libs/graph/src/store/postgres/query/table.rs [hunk 1]
```diff
@@ -399,332 +540,151 @@ fn transpile_json_field(
...
+    Schema,
+}
 
-    pub const fn parameter_type(self) -> ParameterType {
+impl DatabaseColumn for DataTypes {
+    fn parameter_type(self) -> ParameterType {
         match self {
             Self::OntologyId => ParameterType::Uuid,
-            Self::AdditionalMetadata => ParameterType::Object,
+            Self::Schema => ParameterType::Any,
```
Comment (indietyp) [inline apps/hash-graph/libs/graph/src/store/postgres/query/table.rs:578]: why the change from `Object` -> `Any`?

**Review by indietyp — CHANGES_REQUESTED** [review]
mostly minor stylistic comments, and 1-2 questions where I am unsure. Hopefully I didn't miss anything, quite a lot to go through 😅

File: apps/hash-graph/libs/graph/src/store/postgres/query/compile.rs [hunk 2]
```diff
@@ -293,38 +295,38 @@ impl<'p, 'q: 'p, R: PostgresRecord> SelectCompiler<'p, 'q, R> {
...
+                self.statement.distinct.push(stored.column.clone());
                 stored.distinctness = Distinctness::Distinct;
             }
             if stored.ordering.is_none()
                 && let Some((ordering, nulls)) = ordering
             {
                 self.statement
                     .order_by_expression
-                    .push(stored.column, ordering, nulls);
+                    .push(stored.column.clone(), ordering, nulls);
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/postgres/query/compile.rs:306]: Because `Column` (`: Copy`) is now `Expression` (`: !Copy`)

File: apps/hash-graph/libs/graph/src/snapshot/ontology/data_type/batch.rs [hunk 3]
```diff
@@ -1,25 +1,23 @@
...
+        AsClient, InsertionError, PostgresStore,
     },
-    store::{AsClient, InsertionError, PostgresStore},
 };
 
 pub enum DataTypeRowBatch {
     Schema(Vec<DataTypeRow>),
     Relations(HashMap<DataTypeId, Vec<DataTypeRelationAndSubject>>),
-    Embeddings(Vec<DataTypeEmbeddingRow>),
+    Embeddings(Vec<DataTypeEmbeddingRow<'static>>),
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/snapshot/ontology/data_type/batch.rs:20]: Yes, this needs to be owned data only due to the nature of snapshots and how they are processed. I like to unify the snapshot interface with the actual graph interface at some point so I didn't mess with lifetimes here as it wouldn't change anything anyway.

File: apps/hash-graph/libs/graph/src/snapshot/ontology/entity_type/channel.rs [hunk 4]
```diff
@@ -174,10 +175,10 @@ impl Sink<EntityTypeSnapshotRecord> for EntityTypeSender {
         self.schema
             .start_send_unpin(EntityTypeRow {
                 ontology_id,
-                schema: Json(entity_type.schema.clone()),
+                schema: entity_type.schema.clone(),
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/snapshot/ontology/entity_type/channel.rs:178]: No, because the `schema` is also used below for the closed schema, see the comment one line below this

File: apps/hash-graph/libs/graph/src/store/postgres/query/expression/conditional.rs [hunk 5]
```diff
@@ -1,16 +1,23 @@
...
-use crate::store::postgres::query::{AliasedColumn, Transpile, WindowStatement};
+use crate::store::{
+    postgres::query::{
+        table::DatabaseColumn, Alias, AliasedTable, Column, Table, Transpile, WindowStatement,
+    },
+    query::PathToken,
+};
 
-#[derive(Debug, PartialEq, Eq, Hash)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/postgres/query/expression/conditional.rs:10]: This would be a micro-optimization that is not worth spending time on, yet. When this becomes a performance bottleneck I'm quite happy 😄

File: apps/hash-graph/libs/graph/src/store/postgres/query/table.rs (hunk 1 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/postgres/query/table.rs:578]: It didn't change, `git` just didn't calculate the diff propertly

**Review by indietyp — DISMISSED** [review]
LGTM! 🎉

**Review by indietyp — DISMISSED** [review]

**Review by indietyp — DISMISSED** [review]

**Review by indietyp — APPROVED** [review]

**Review by indietyp — APPROVED** [review]
